autopush-common/src/reliability.rs

```diff
@@ -179,33 +169,59 @@ impl PushReliability {
     ) -> Result<()> {
-        let purged: Vec<String> = conn.zrangebyscore(EXPIRY, 0, expr as isize)?;
+        // First, get the list of values that are to be purged.
+        let mut purge_pipe = redis::Pipeline::new();
+        purge_pipe.zrangebyscore(EXPIRY, 0, expr as isize);
+        let result_list: Vec<Vec<redis::Value>> = purge_pipe.query_async(conn).await?;
+
+        // Now purge each of the values by resetting the counts and removing the item from the purge set.
         let mut pipeline = redis::Pipeline::new();
-        for key in purged {
-            let Some((state, _id)) = key.split_once('#') else {
-                let err = "Invalid key stored in Reliability datastore";
-                error!("🔍🟥 {} [{:?}]", &err, &key);
-                return Err(ApcErrorKind::GeneralError(err.to_owned()).into());
-            };
-            pipeline.hincr(COUNTS, state, -1);
-            pipeline.zrem(EXPIRY, key);
+        for result_values in result_list {
+            for purged in result_values {
+                let key = redis::from_redis_value::<String>(&purged)?;
+                let Some((state, _id)) = key.split_once('#') else {
+                    let err = "Invalid key stored in Reliability datastore";
+                    error!("🔍🟥 {} [{:?}]", &err, &key);
+                    return Err(ApcErrorKind::GeneralError(err.to_owned()).into());
+                };
+                pipeline.hincr(COUNTS, state, -1);
+                pipeline.zrem(EXPIRY, key);
+            }
         }
-        pipeline.exec(conn)?;
+        pipeline.exec_async(conn).await?;
         Ok(())
     }
 
     // Return a snapshot of milestone states
     // This will probably not be called directly, but useful for debugging.
     pub async fn report(&self) -> Result<HashMap<String, i32>> {
-        if let Some(client) = &self.client {
-            if let Ok(mut conn) = client.get_connection() {
-                return Ok(conn.hgetall(COUNTS).map_err(|e| {
+        if let Some(pool) = &self.pool {
+            if let Ok(mut conn) = pool.get().await {
+                return Ok(conn.hgetall(COUNTS).await.map_err(|e| {
                     ApcErrorKind::GeneralError(format!("Could not read report {:?}", e))
                 })?);
             }
         }
         Ok(HashMap::new())
     }
+
+    pub async fn health_check<'a>(&self) -> Result<&'a str> {
+        if let Some(pool) = &self.pool {
+            let mut conn = pool.get().await.map_err(|e| {
+                ApcErrorKind::GeneralError(format!(
+                    "Could not connect to reliability datastore: {:?}",
+                    e,
+                ))
+            })?;
+            // Add a type here, even though we're tossing the value, in order to prevent the `FromRedisValue` warning.
+            conn.ping::<()>().await.map_err(|e| {
+                ApcErrorKind::GeneralError(format!("Could not ping reliability datastore: {:?}", e))
+            })?;
+            Ok("up")
```

Comment:
we're already doing "OK"/"ERROR" for the other db status, we should probably keep it consistent?